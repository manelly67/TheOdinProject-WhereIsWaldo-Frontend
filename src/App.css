body {
  color: var(--color-base-text);
  background-color: var(--color-bg-first);
}
 

#root {
  max-width: 1280px;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
  color: var(--color-base-text);
  background-color: var(--color-bg-first);
}

.setOfButtons {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  gap: 10px;
}

.iconImg {
  height: 30px;
  width: auto;
}


@media (max-width: 770px) {
 .bar {
  position: absolute;
  top: 4%;
  width: 300px;
  display: grid;
  grid-template-rows: 40px;
  grid-template-columns: 1fr 1fr 1fr 1fr 2fr;

 }
 .themeButton{
  grid-row: 1;
  grid-column-start: 5;
  grid-column-end: 6;
 }

.setOfLinks{
  grid-row: 1;
  grid-column-start: 1;
  grid-column-end:3;
  }

}


.setOfLinks {
  position: absolute;
  top: 15%;
  right: 3%;

  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  gap: 3px;
}

button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: var(--color-bg-button);
  cursor: pointer;
  transition: border-color 0.25s;
  transition: transform 250ms ease-in-out;
}
button:hover {
  border-color: #84cc16;
  transform: scale(1.02);
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}


.themeButton {
  position: absolute;
  top: 4%;
  right: 3%;
  z-index: 1;
  padding: 0.3em 0.9em;

  transition: transform 250ms ease-in-out;
}

.themeButton:hover {
  transform: scale(1.1);
 
}

.themeButton:active {
  transform: scale(1.1);
 
}
