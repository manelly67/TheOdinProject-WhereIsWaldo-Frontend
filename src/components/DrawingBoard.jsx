import { Link } from "react-router-dom";
import { useState, useEffect, useCallback } from "react";
import { useLocation } from "react-router-dom";
import ToggleTheme from "./ToggleTheme";
import TargetingBox from "./TargetinBox";
import DropdownMenu from "./DropdownMenu";
import TagTheChar from "./TagTheChar";
import Message from "./Message";
import AskForName from "./AskForName";
import styles from "../styles/Box.module.css";
import { moveToCoord, clickInsideImg } from "./coord";
import { urlAddresses } from "../assets/urlAdresses";

let didInitImg1 = false;
let didInitImg2 = false;

const DrawingBoard = () => {
  const { box, placeMenu, dropdownMenu, visible } = styles;
  const [gameName, setGameName] = useState(null);
  const titleDiv = document.querySelector("title");
  if (titleDiv) {
    titleDiv.textContent = gameName;
  }
  const location = useLocation();

  const [game, setGame] = useState(null);
  const [messageObj, setMessageObj] = useState(null);
  const [isInTop, setIsInTop] = useState(false);
  const [dontAskAgain, setDontAskAgain] = useState(false);
  const imgSource = game === null ? null : game.picture.src_image;
  const imgId = game === null ? null : game.picture.id_image;
  const imgCharacters = game === null ? null : game.picture.characters;
  const score = game === null ? null : formatScore(game.timeRecord);
  
  const width = window.innerWidth;
  const height = window.innerHeight;
  const [coords, setCoords] = useState({ x: 0, y: 0 });
  const [endcoords, setEndcoords] = useState({ x: 0, y: 0 });
  const [W, setW] = useState(0);
  const [H, setH] = useState(0);
  const [tagginCoords, setTagginCoords] = useState({ x: 0, y: 0 });
  const [clickImg, setClickImg] = useState(false);
  const [stop, setStop] = useState(false);

  const getData = useCallback(async (url) => {
    try {
      await fetch(url, {
        method: "POST",
        credentials: "same-origin",
        headers: {
          "Content-Type": "application/json",
          Connection: "keep-alive",
        },
      })
        .then((res) => res.json())
        .then((data) => {
          if (data.game) {
            const { game } = data;
            setGame(game);
          }
        })
        .catch((err) => {
          console.log(err);
        });
    } catch (error) {
      alert("Something was wrong. try again later");
      console.log(error);
    }
  }, []);

  const initBoard = useCallback(async () => {
  
    if (location.state !== null) {
      const { player, gameName } = location.state;
      setGameName(gameName);

      if (gameName !== null) {
        if (gameName === "Waldo In The Galactic City") {
          const url = `${urlAddresses.createGame}/img-1/${player.id}`;
          if (!didInitImg1) {
            didInitImg1 = true;
            await getData(url);
          }
        }
        if (gameName === "Oh! Waldo is not here") {
          const url = `${urlAddresses.createGame}/img-2/${player.id}`;
          if (!didInitImg2) {
            didInitImg2 = true;
            await getData(url);
          }
        }
      }
    }
  }, [location.state, getData]);

  const getImgCoord = useCallback(() => {
    const imageRef = document.getElementById(`${gameName}`);
    if (imageRef) {
      const rect = imageRef.getBoundingClientRect();
      setCoords({
        x: Number(rect.left.toFixed(10)),
        y: Number(rect.top.toFixed(10)),
      });
      setEndcoords({
        x: Number(rect.right.toFixed(10)),
        y: Number(rect.bottom.toFixed(10)),
      });
      setW(Number(rect["width"].toFixed(4)));
      setH(Number(rect["height"].toFixed(4)));
    }
  }, [gameName]);

  const updateData = useCallback(async (url) => {
    try {
      const response = await fetch(url, { method: "GET" });
      const responseData = await response.json();
      if (responseData.game) {
        const { game } = responseData;
        setGame(game);
      }
    } catch (error) {
      alert("Something was wrong. try again later");
      console.log(error);
    }
  }, []);

  const checkIsInTopTen = useCallback(async (game) => {
    const url = `${urlAddresses.topTEN}/${game.picture.id_image}`;
    try {
      const response = await fetch(url);
      const responseData = await response.json();
      if (responseData.top10) {
        const { top10 } = responseData;
        top10.forEach((r) => {
          if (r) {
            if (r.id === game.game_id) {
              setIsInTop(true);
            }
          }
        });
      }
    } catch (error) {
      alert("Something was wrong. try again later");
      console.log(error);
    }
  }, []);

  useEffect(() => {
    if (game === null) {
      initBoard();
    }
  }, [game, initBoard]);

  useEffect(() => {
    if (location.state !== null) {
      const { player, gameName } = location.state;
      if (gameName !== null && player !== null) {
        if (gameName === "Waldo In The Galactic City") {
          const url = `${urlAddresses.createGame}/img-1/${player.id}`;
          updateData(url);
        }
        if (gameName === "Oh! Waldo is not here") {
          const url = `${urlAddresses.createGame}/img-2/${player.id}`;
          updateData(url);
        }
      }
    }
  }, [location.state, updateData]);

  useEffect(() => {
    if (imgSource !== null) {
      getImgCoord();
    }
  }, [gameName, imgSource, getImgCoord]);

  useEffect(() => {
    const temp = clickInsideImg(tagginCoords, coords, endcoords);
     setClickImg(temp);
  }, [tagginCoords, coords, endcoords]);

  useEffect(() => {
    const f = document.getElementById("box");
    if (f) {
      if (tagginCoords.x === 0 && tagginCoords.y === 0) {
        f.style.transform = `translateY(${tagginCoords.y}px)`;
        f.style.transform += `translateX(${tagginCoords.x}px)`;
      } else {
        f.style.transform = `translateY(${tagginCoords.y - 5}px)`;
        f.style.transform += `translateX(${tagginCoords.x - 15}px)`;
      }
    }
  }, [tagginCoords]);

  useEffect(() => {
    switch (game === null) {
      case true:
        break;
      case false:
        switch (game.status === "ENDED") {
          case true:
            setStop(true);
            if (game.player.name_player === "ANONIMOUS") {
              if (!dontAskAgain) {
                checkIsInTopTen(game);
              }
            }
            break;
          case false:
            switch (game.status === "GAMING") {
              case true: {
                document.addEventListener(
                  "click",
                  (ev) => {
                    let { movetoX, movetoY } = moveToCoord(
                      ev.clientX,
                      ev.clientY,
                      ev.pageX,
                      ev.pageY
                    );
                    setTagginCoords({ x: movetoX, y: movetoY });
                  },
                  false
                );
                return () => {
                  document.removeEventListener(
                    "click",
                    (ev) => {
                      let { movetoX, movetoY } = moveToCoord(
                        ev.clientX,
                        ev.clientY,
                        ev.pageX,
                        ev.pageY
                      );
                      setTagginCoords({ x: movetoX, y: movetoY });
                    },
                    false
                  );
                };
              }
              default:
                break;
            }
        }
    }
  }, [game, dontAskAgain, checkIsInTopTen]);

  function formatScore(score) {
    if (score) {
      const { days, hours, minutes, seconds } = score;
      let d = days ? `${days} days` : "";
      let h = hours ? `${hours} hours` : "";
      let m = minutes ? `${minutes} minutes` : "";
      let s = seconds ? `${seconds} seconds` : "";
      return `${d} ${h} ${m} ${s}`;
    }
  }

  return (
    <>
      <div className="bar">
        <>
          {!stop ? (
            <>
              <div id="box" className={`${box}`}>
                <div>
                  <TargetingBox />
                </div>
                <div>
                  <DropdownMenu
                    clickImg={clickImg}
                    setClickImg={setClickImg}
                    game={game}
                    setGame={setGame}
                    coords={coords}
                    W={W}
                    H={H}
                    tagginCoords={tagginCoords}
                    setTagginCoords={setTagginCoords}
                    placeMenu={placeMenu}
                    dropdownMenu={dropdownMenu}
                    visible={visible}
                    imgCharacters={imgCharacters}
                    setMessageObj={setMessageObj}
                  />
                </div>
              </div>
            </>
          ) : null}

          <div>
            <ToggleTheme theme="dark" />
          </div>

          <div className="setOfLinks">
            <Link to="/" style={{ fontSize: "1.3rem" }}>
              HOME
            </Link>
            <Link
              to={"/top_ten"}
              state={{ img_id: imgId, gameName: gameName }}
              style={{ fontSize: "1.3rem" }}
            >
              TOP TEN
            </Link>
            {!game ? null : !game.timeInSeconds ? null : (
              <>
                <div>
                  <p> Your SCORE:</p>
                  <p style={{ width: "80px" }}>{score}</p>
                </div>
              </>
            )}
          </div>
        </>
      </div>
      <div>
        {!imgSource ? (
          <>
            <div>
              <p>loading</p>
              <p>
                Dear user, this is a study project, please wait 1 minute for the
                server to wake up.
              </p>
            </div>
          </>
        ) : (
          <>
            <img
              id={gameName}
              src={imgSource}
              alt={gameName}
              className=""
              width={width * 0.75}
              height={
                width * 0.8 * 0.65 < height ? width * 0.8 * 0.65 : height * 0.95
              }
            ></img>
          </>
        )}
      </div>
      {imgCharacters ? (
        <TagTheChar imgCharacters={imgCharacters} coords={coords} W={W} H={H} />
      ) : null}

      {messageObj ? (
        <Message messageObj={messageObj} setMessageObj={setMessageObj} />
      ) : null}

      {isInTop ? (
        <AskForName
          game={game}
          setIsInTop={setIsInTop}
          setDontAskAgain={setDontAskAgain}
        />
      ) : null}
    </>
  );
};

export default DrawingBoard;
