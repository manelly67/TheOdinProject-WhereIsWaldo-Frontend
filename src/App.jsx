import { useState, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import "./styles/App.css";
import ToggleTheme from "./components/ToggleTheme";

let didInit = false;

function App() {
  const titleDiv = document.querySelector("title");
  if (titleDiv) {
    titleDiv.textContent = "WHERE IS WALDO";
  }

  const navigate = useNavigate();
  const name_game_1 = "Waldo In The Galactic City";
  const name_game_2 = "Oh! Waldo is not here";
  const [playerId, setPlayerId] = useState(null);
  const [playerName, setPlayerName] = useState(null);
  const player = playerId === null ? null : { id: playerId, name: playerName };

  useEffect(() => {
    if (!didInit) {
      didInit = true;
      // âœ… Only runs once per app load
      const id = crypto.randomUUID();
      setPlayerId(id);
      setPlayerName("anonymous");
    }
  }, []);

  console.log(player);

  return (
    <>
      <ToggleTheme theme="light" />
      <h1>Where is Waldo - The Game</h1>

      <div className="setOfButtons">
        <button
          onClick={() => {
            navigate("/board", {
              replace: true,
              state: { player: player, gameName: name_game_1 },
            });
          }}
        >
          <h3>{name_game_1}</h3>
        </button>
        <button
          onClick={() => {
            navigate("/board", {
              replace: true,
              state: { player: player, gameName: name_game_2 },
            });
          }}
        >
          <h3>{name_game_2}</h3>
        </button>
      </div>
    </>
  );
}

export default App;
