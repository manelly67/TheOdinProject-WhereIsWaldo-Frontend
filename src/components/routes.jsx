import App from "../App";
import ErrorPage from "./Error_page";
import DrawingBoard from "./DrawingBoard";
import TopTen from "./TopTen";

const routes = [
  {
    index: true,
    path: "/",
    element: <App />,

    errorElement: <ErrorPage />,
  },
  {
    path: "board",
    element: <DrawingBoard />,
  },
  {
    path: "top_ten",
    element: <TopTen />,
  },
];

export default routes;
