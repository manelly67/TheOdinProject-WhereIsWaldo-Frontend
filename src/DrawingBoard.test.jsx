import React from "react";
import { describe, it, expect, vi } from "vitest";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter, createBrowserRouter, RouterProvider } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import routes from "./components/routes.jsx";
import App from "./App";
import DrawingBoard from "./components/DrawingBoard";


describe("simple use of mock fuction", () => {
    it("mock function", () => {
      const fn = vi.fn();
      fn("hello world");
  
      expect(fn.mock.calls[0]).toEqual(["hello world"]);
  
      const market = {
        getApples: () => 100,
      };
  
      const getApplesSpy = vi.spyOn(market, "getApples");
      market.getApples();
  
      expect(getApplesSpy.mock.calls.length).toEqual(1);
      expect(market.getApples()).toBe(100);
    });
  });
  
  describe("display DrawingBoard", () => {
    
    it("display the page when location.state is null", () => {
        const { container } = render(
            <MemoryRouter initialEntries={["/board"]}>
              <DrawingBoard />
            </MemoryRouter>
          );
          expect(container).toMatchSnapshot();
          expect(screen.getByText(/Dear user, this is a study project, please wait 1 minute for the server to wake up./i)).toBeInTheDocument();
    });

    it("display drawing board when location.state is not null", async () => {
        const router = createBrowserRouter(routes);
        render(<RouterProvider router={router} />);
        const user = userEvent.setup();
    
        const buttons = screen.getAllByRole("button");
        const button = buttons[1];
    
        await user.click(button);
    
        expect(
          screen.getByAltText(/Waldo In The Galactic City/i)
        ).toBeInTheDocument();

       

      });



  });