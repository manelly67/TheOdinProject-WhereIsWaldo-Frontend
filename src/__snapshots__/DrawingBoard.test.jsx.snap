// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`display DrawingBoard > display the page when location.state is null 1`] = `
<div>
  <div
    class="bar"
  >
    <div>
      <button
        class="themeButton"
      >
        <img
          alt="theme-light-dark"
          class="iconImg"
          height="30px"
          src="/src/assets/theme-light-dark.png"
          width="30px"
        />
      </button>
    </div>
    <div
      class="setOfLinks"
    >
      <a
        data-discover="true"
        href="/"
        style="font-size: 1.3rem;"
      >
        HOME
      </a>
    </div>
  </div>
  <div>
    <div>
      <p>
        loading
      </p>
      <p>
        Dear user, this is a study project, please wait 1 minute for the server to wake up.
      </p>
    </div>
  </div>
</div>
`;
